engine/communication/src/behaviour/mod.rs

```diff
@@ -0,0 +1,256 @@
+    /// own ID in the DHT. This introduces the local node to the other nodes
+    /// in the DHT and populates its routing table with the closest neighbours.
+    ///
+    /// Subsequently, all buckets farther from the bucket of the closest neighbour are
+    /// refreshed by initiating an additional bootstrapping query for each such
+    /// bucket with random keys.
+    ///
+    /// Returns `Ok` if bootstrapping has been initiated with a self-lookup, providing the
+    /// `QueryId` for the entire bootstrapping process. The progress of bootstrapping is
+    /// reported via [`KademliaEvent::QueryResult{QueryResult::Bootstrap}`] events,
+    /// with one such event per bootstrapping query.
+    ///
+    /// Returns `Err` if bootstrapping is impossible due an empty routing table.
+    ///
+    /// > **Note**: Bootstrapping requires at least one node of the DHT to be known.
+    /// > See libp2p [`Kademlia::add_address`].
+    #[cfg(feature = "kademlia")]
+    fn kad_bootstrap(&mut self) -> QueryResult<QueryId> {
+        self.kademlia
+            .bootstrap()
+            .map_err(|e| QueryError::KademliaError(format!("Could not bootstrap:{:?}", e.to_string())))
+    }
+}
+
+impl<C: InboundEventCodec + Send + 'static> P2PNetworkBehaviour<C> {
+    /// Creates a new P2PNetworkbehaviour that defines the communication with the libp2p swarm.
+    /// It combines the following protocols from libp2p:
+    /// - mDNS for peer discovery within the local network
+    /// - RequestResponse Protocol for sending request and reponse messages. This stronghold-communication library
+    ///   defines a custom version of this protocol that for sending pings, string-messages and key-value-records.
+    /// - kademlia (if the "kademlia"-feature is enabled) for managing peer in kademlia buckets and publishing/ reading records
+    ///
+    /// # Example
+    /// ```no_run
+    /// use communication::{
+    ///     behaviour::{InboundEventCodec, P2PNetworkBehaviour, SwarmContext},
+    ///     error::QueryResult,
+    ///     message::{Request, Response},
+    /// };
+    /// use libp2p::{
+    ///     kad::KademliaEvent,
+    ///     core::{identity::Keypair, Multiaddr, PeerId},
+    ///     request_response::{RequestId, ResponseChannel, RequestResponseEvent},
+    /// };
+    ///
+    /// let local_keys = Keypair::generate_ed25519();
+    ///
+    /// struct Handler();
+    /// impl InboundEventCodec for Handler {
+    ///    fn handle_request_response_event(
+    ///       _swarm: &mut impl SwarmContext,
+    ///       _event: RequestResponseEvent<Request, Response>,
+    ///    ) {}
+    ///
+    ///    fn handle_kademlia_event(_swarm: &mut impl SwarmContext, _result: KademliaEvent) {}
+    /// }
+    ///
+    /// let behaviour = P2PNetworkBehaviour::<Handler>::new(local_keys.public()).unwrap();
+    /// ```
+    pub fn new(_public_key: PublicKey) -> QueryResult<Self> {
+        let peer_id = PeerId::from(_public_key);
```

Comment:
should remove the underscore.  Only really exists if you don't plan to use the argument which you are using here.